deflect/qt/QuickRenderer.cpp

```diff
@@ -0,0 +1,201 @@
+}
+
+void QuickRenderer::render()
+{
+    QMutexLocker lock( &_mutex );
+    QCoreApplication::postEvent( this, new QEvent( QEvent::User ));
+
+    // the main thread has to be blocked for sync()
+    _cond.wait( &_mutex );
+}
+
+bool QuickRenderer::event( QEvent* e )
+{
+    if( e->type() == QEvent::User )
+    {
+        _onRender();
+        return true;
+    }
+    return QObject::event( e );
+}
+
+void QuickRenderer::_onInit()
+{
+    // Qt Quick may need a depth and stencil buffer
+    QSurfaceFormat format_;
+    format_.setDepthBufferSize( 16 );
+    format_.setStencilBufferSize( 8 );
+
+    _context = new QOpenGLContext;
+    _context->setFormat( format_ );
+    _context->create();
+
+    // Test if user has setup shared GL contexts (QtWebEngine::initialize).
+    // If so, setup global share context needed by the Qml WebEngineView.
+    if( QCoreApplication::testAttribute( Qt::AA_ShareOpenGLContexts ))
+#if DEFLECT_USE_QT5GUI
+        qt_gl_set_global_share_context( _context );
+#else
+        qWarning() << "DeflectQt was not compiled with WebEngineView support";
+#endif
+
+    if( _offscreen )
+    {
+        _offscreenSurface = new QOffscreenSurface;
+        // Pass _context->format(), not format_. Format does not specify and color
+        // buffer sizes, while the context, that has just been created, reports a
+        // format that has these values filled in. Pass this to the offscreen
+        // surface to make sure it will be compatible with the context's
+        // configuration.
+        _offscreenSurface->setFormat( _context->format( ));
+        _offscreenSurface->create();
+    }
+
+    _context->makeCurrent( _getSurface( ));
+    _renderControl.initialize( _context );
+    _initialized = true;
+}
+
+void QuickRenderer::_onStop()
+{
+    _context->makeCurrent( _getSurface( ));
+
+    _renderControl.invalidate();
+
+    delete _fbo;
+    _fbo = nullptr;
+
+    _context->doneCurrent();
+
+    delete _offscreenSurface;
+    _offscreenSurface = nullptr;
+    delete _context;
+    _context = nullptr;
+
+    emit afterStop();
+}
+
+void QuickRenderer::_ensureFBO()
+{
+    if( !_offscreen )
+        return;
+
+    if( _fbo &&
+        _fbo->size() != _quickWindow.size() * _quickWindow.devicePixelRatio())
+    {
+        delete _fbo;
+        _fbo = nullptr;
+    }
+
+    if( !_fbo )
+    {
+        _fbo = new QOpenGLFramebufferObject(
+                    _quickWindow.size() * _quickWindow.devicePixelRatio(),
+                    QOpenGLFramebufferObject::CombinedDepthStencil );
+        _quickWindow.setRenderTarget( _fbo );
+    }
+}
+
+QSurface* QuickRenderer::_getSurface()
+{
+    if( _offscreen )
+        return _offscreenSurface;
+    return &_quickWindow;
+}
+
+void QuickRenderer::_onRender()
+{
+    if( !_initialized )
+        return;
+
+    {
+        QMutexLocker lock( &_mutex );
+
+        _context->makeCurrent( _getSurface( ));
+
+        _ensureFBO();
```

Comment:
I would find it clearer to do:
```
if( offscreen )
    _ensureFBO();
```
instead of:
```
void QuickRenderer::_ensureFBO()
if( !_offscreen )
    return;
```